From 781b9ca56524e0815c48f928a8a2c6a4ae2c3f0e Mon Sep 17 00:00:00 2001
From: Tobias Schramm <t.schramm@manjaro.org>
Date: Thu, 13 May 2021 15:50:50 +0200
Subject: [PATCH 12/20] mmc: pwrseq_simple: add feature to cut power on
 shutdown

Some SDIO modules tend to be problematic when reinitialized after a reboot
without cutting their power supply. One example of this behaviour is the
Ampak AP6255 WiFi module, which will not enumerate if reinitialized
without cutting the power to it before reinitialization.
In theory this should not be an issue since there is a reset preformed
before power up by this driver already. However, unfortunately many of
those WiFi modules are also Bluetooth modules. Thus the also have two
power enable pins, one for WiFi and another for Bluetooth.
Unfortunately both those signals are or'-ed internally and control the
same regulators. Thus it is only possible to reset the WiFi section of the
module if both, WiFi enable and Bluetooth enable are deasserted together.
Thus this patch is mirrored in the relevant HCI dirvers.

Signed-off-by: Tobias Schramm <t.schramm@manjaro.org>
---
 .../bindings/mmc/mmc-pwrseq-simple.yaml       |  4 ++++
 drivers/mmc/core/pwrseq_simple.c              | 24 ++++++++++++++++---
 2 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/Documentation/devicetree/bindings/mmc/mmc-pwrseq-simple.yaml b/Documentation/devicetree/bindings/mmc/mmc-pwrseq-simple.yaml
index 6cd57863c1db..9ffb95889a58 100644
--- a/Documentation/devicetree/bindings/mmc/mmc-pwrseq-simple.yaml
+++ b/Documentation/devicetree/bindings/mmc/mmc-pwrseq-simple.yaml
@@ -49,6 +49,10 @@ properties:
       during power off of the card.
     $ref: /schemas/types.yaml#/definitions/uint32
 
+  power-off-on-shutdown:
+    description:
+      Force assertion of reset-gpios (if any) on shutdown
+
 required:
   - compatible
 
diff --git a/drivers/mmc/core/pwrseq_simple.c b/drivers/mmc/core/pwrseq_simple.c
index ea4d3670560e..fb4f8a6bf0fa 100644
--- a/drivers/mmc/core/pwrseq_simple.c
+++ b/drivers/mmc/core/pwrseq_simple.c
@@ -29,6 +29,7 @@ struct mmc_pwrseq_simple {
 	u32 power_off_delay_us;
 	struct clk *ext_clk;
 	struct gpio_descs *reset_gpios;
+	bool do_shutdown;
 };
 
 #define to_pwrseq_simple(p) container_of(p, struct mmc_pwrseq_simple, pwrseq)
@@ -80,10 +81,8 @@ static void mmc_pwrseq_simple_post_power_on(struct mmc_host *host)
 		msleep(pwrseq->post_power_on_delay_ms);
 }
 
-static void mmc_pwrseq_simple_power_off(struct mmc_host *host)
+static void mmc_pwrseq_simple_power_off_(struct mmc_pwrseq_simple *pwrseq)
 {
-	struct mmc_pwrseq_simple *pwrseq = to_pwrseq_simple(host->pwrseq);
-
 	mmc_pwrseq_simple_set_gpios_value(pwrseq, 1);
 
 	if (pwrseq->power_off_delay_us)
@@ -96,6 +95,13 @@ static void mmc_pwrseq_simple_power_off(struct mmc_host *host)
 	}
 }
 
+static void mmc_pwrseq_simple_power_off(struct mmc_host *host)
+{
+	struct mmc_pwrseq_simple *pwrseq = to_pwrseq_simple(host->pwrseq);
+
+	mmc_pwrseq_simple_power_off_(pwrseq);
+}
+
 static const struct mmc_pwrseq_ops mmc_pwrseq_simple_ops = {
 	.pre_power_on = mmc_pwrseq_simple_pre_power_on,
 	.post_power_on = mmc_pwrseq_simple_post_power_on,
@@ -134,6 +140,9 @@ static int mmc_pwrseq_simple_probe(struct platform_device *pdev)
 	device_property_read_u32(dev, "power-off-delay-us",
 				 &pwrseq->power_off_delay_us);
 
+	pwrseq->do_shutdown =
+		device_property_read_bool(dev, "power-off-on-shutdown");
+
 	pwrseq->pwrseq.dev = dev;
 	pwrseq->pwrseq.ops = &mmc_pwrseq_simple_ops;
 	pwrseq->pwrseq.owner = THIS_MODULE;
@@ -151,9 +160,18 @@ static int mmc_pwrseq_simple_remove(struct platform_device *pdev)
 	return 0;
 }
 
+static void mmc_pwrseq_simple_shutdown(struct platform_device *pdev)
+{
+	struct mmc_pwrseq_simple *pwrseq = platform_get_drvdata(pdev);
+
+	if (pwrseq->do_shutdown)
+		mmc_pwrseq_simple_power_off_(pwrseq);
+}
+
 static struct platform_driver mmc_pwrseq_simple_driver = {
 	.probe = mmc_pwrseq_simple_probe,
 	.remove = mmc_pwrseq_simple_remove,
+	.shutdown = mmc_pwrseq_simple_shutdown,
 	.driver = {
 		.name = "pwrseq_simple",
 		.of_match_table = mmc_pwrseq_simple_of_match,
-- 
2.31.1

